rfcs/0010-mempool-dos-mitigation.md

```diff
@@ -0,0 +1,302 @@
+## Mitigations for DOS attacks based on bogus transactions
+[summary]: #summary
+
+We add some rules for deciding whether to store incoming transactions in the
+mempool and gossip them. FIXME details.
+
+
+## Motivation
+
+[motivation]: #motivation
+
+Every transaction consumes resources: computation, storage, and bandwidth. In
+the case of transactions that are eventually included in a block, those
+resources are priced by transaction fees. But transactions that are never
+included in a block don't pay transaction fees, so without good design bogus
+transactions could be a vector for cheap DoS attacks. We want to make sure it
+costs enough to make nodes consume resources that a DoS attack is too expensive
+to be worthwhile.
+
+(The transaction fees for *mined* transactions may or may not accurately reflect
+the total cost to all network participants for processing the transaction, but
+that's a problem for another day.)
+
+We can't control what transactions we receive, but we can control whether we
+store them and whether we forward them on, making other nodes deal with them.
+
+## Detailed design
+
+[detailed-design]: #detailed-design
+
+As a general principle, we want to accept transactions into the mempool iff they
+will eventually be mined into a future block. That's the purpose of the mempool.
+For DoS protection, we assume that the cost of including a transaction in a
+block is sufficient to deter DoS attacks on proposers, since the cost of
+creating a SNARK greatly exceeds the cost of checking and storing an incoming
+transaction. Where possible, we prefer to charge for things rather than banning
+them. We don't mind if someone uses a lot of resources, so long as those
+resources are paid for at a price that network participants would be happy with.
+
+### When we first recieve a transaction: rules
+
+When we receive a gossiped transaction, we will check the below constraints. If
+any of the checks fail, we ignore the transaction and do not gossip it.
+
+* Validity:
+  1. The signature.
+  2. The sender's account exists.
+  3. The sender's balance (inclusive of any pending txs from the same account
+  with lower nonce) is >= the transaction amount + fee.
+  4. The tx nonce is equal to the sending account's next nonce, inclusive of
+  transactions already in the mempool. (If it conflicts with one in the mempool,
+  see below.)
+* Mining probability:
+  1. The (adjusted?) fee is >= the lowest fee paid in a transaction included in
+     the last `min_fee_window` blocks, times `min_fee_discount_factor`, a value
+     in the interval (0,1). FIXME: This minimum is attacker controlled. An
+     attacker who wins at least one slot and proves one SNARK can set it to
+     zero. Can't use a (exponentially weighted) moving average either, for the
+     opposite issue. Maybe don't do a set minimum and rely on the fee rules for
+     when the pool is full?
+
+If we get any transactions where the signature is invalid, we blacklist the
```

Comment:
Is blacklisting forever? What if the invalid signature was caused by a gamma ray striking the sender's machine?